[Logs UI] Implement log stream page state as a state machine

weltenwort opened the pull request:
## :memo: Summary

This takes the first step in reimplementing the Logs UI's stream page state management using XState state charts.

closes #145131

## :female_detective: Review notes

### Package-like directory structure

We intended to write most of the newly added code as `packages` in order to not increase the burden when migrating to a mostly-packages architecture in the future. But since we depend on several types and values that are themselves not yet packaged in this way, we couldn't follow through with this plan. Instead we decided to put most of the new state management code into a new directory `infra/public/observability_logs` in a way that mimics the future package structure. That way we hope to make the migration to real packages such as `packages/observability-logs/log_view_state` as seamless as possible once all dependencies used in those files have been packaged.

### Introduction of XState

This PR adds [XState](https://xstate.js.org/) as a dependency. It allows us to express the life-cycle of state and behavior as state charts in a robust and self-documenting way. The patterns and development tools we're using are described in [infra/docs/state_machines/xstate_usage.md](https://github.com/weltenwort/kibana/blob/log-stream-state-machine/x-pack/plugins/infra/docs/state_machines/xstate_usage.md), which makes that file a good place to start the review. The [XState docs](https://xstate.js.org/docs/) offer both a reference but also a basic [introduction to state machines and state charts](https://xstate.js.org/docs/guides/introduction-to-state-machines-and-statecharts/#states). The ["Welcome to the world of Statecharts"](https://statecharts.dev/) page is another well-written introduction to state charts in UI development.

### Logs UI stream page changes

The log stream page's state is now managed using an XState state machine:

![image](https://user-images.githubusercontent.com/973741/204802241-cc1ef36e-3c39-403f-a86c-3ad77b0a039d.png)

It communicates with the log view state machine described below to react to log view status changes.

### `LogView` changes

The `useLogView` hook is now backed by an XState state machine. To reduce code churn the signature of the hook has been kept mostly identical. New usages should consider accessing the state machine instead

![image](https://user-images.githubusercontent.com/973741/204865776-ee392316-d7b5-477c-bf93-1cf1352125fb.png)

## :crystal_ball: Follow-up work

- https://github.com/elastic/kibana/issues/142764
- https://github.com/elastic/kibana/issues/145133
- https://github.com/elastic/kibana/issues/145137

elasticmachine commented:
Pinging @elastic/infra-monitoring-ui (Team:Infra Monitoring UI)

tonyghiani commented on x-pack/plugins/infra/public/hooks/use_log_view.ts:
```diff
@@ -5,118 +5,126 @@
-  const [logViewStatus, setLogViewStatus] = useState<LogViewStatus | undefined>(undefined);
-
-  const [loadLogViewRequest, loadLogView] = useTrackedPromise(
+  const [logViewStateNotifications] = useState(() => createLogViewNotificationChannel());
```
**nit**: if this is the case whether `logViewStateNotifications` is not a reactive value (it never triggers a rerender), using a `useRef` call may be more suitable for this use case:
```js
const logViewStateNotificationsRef = useRef(createLogViewNotificationChannel())
```
If the `createLogViewNotificationChannel` is an expensive computation and we want to still lazy load the initial value that we'll never update again, we also have two possible approaches using `useRef` or `useMemo`:
```js
const logViewStateNotificationsRef = useRef(null)
if (logViewStateNotificationsRef.current === null) {
  logViewStateNotificationsRef.current = createLogViewNotificationChannel()
}

// or with useMemo
const logViewStateNotifications = useMemo(createLogViewNotificationChannel, [])
```
The expected result is identical to using `useState`, the only difference is we stop treating this value as a reactive one.

weltenwort replied:
> if the `createLogViewNotificationChannel` is an expensive computation and we want to still lazy load the initial value that we'll never update again, we also have two possible approaches using `useRef` or `useMemo`

It creates an observable, several functions, and objects so I would say it's too expensive to unnecessarily throw away on each render.

Overall, can you explain why you don't agree with this use of `useState`? The single instance of the notification channel is "state" with the special property that we currently don't have the need to replace it.

The semantics of `useRef` and `useMemo` actually fit worse IMHO, because

- `useRef` creates a container for a value that can change over time without the identity of the container changing. This is not what we need here.
- `useMemo` doesn't make any guarantees about when and how often the value is calculated. The docs explicitly state that one should not rely on that for referential stability.

So `useState` with an initializer, but without a `setState` is a really good match to implement a value whose identity is tied to the lifetime of the component.

Does that make sense?

tonyghiani commented on x-pack/plugins/infra/public/hooks/use_log_view.ts:
```diff
@@ -5,118 +5,126 @@
-    .filter(isRejectedPromiseState)
-    .map(({ value }) => (value instanceof Error ? value : new Error(`${value}`)));
+  const latestLoadLogViewFailures = useSelector(logViewStateService, (state) =>
+    state.matches('loadingFailed') ||
+    state.matches('resolutionFailed') ||
+    state.matches('checkingStatusFailed')
```
**nit**: when matching against multiple states, the [state documentation](https://xstate.js.org/docs/guides/states.html#state-methods-and-properties) suggest using the `.some` iterator on a states array:
```js
['loadingFailed', 'resolutionFailed', 'checkingStatusFailed'].some(state.matches)
```
It may be useful to improve readability

weltenwort replied:
I chose that approach to improve type safety. If we were to rename a state, the compiler would call that out in the current implementation, but not in the `Array.some` variant. Adding type annotations to the array to ensure the compiler knows only state names are allowed here would probably be even less readable than unrolling the boolean disjunction.

In addition (in other cases, but not this specific selector) the `Array.some()` approach doesn't cause the compiler to narrow the typestate (and therefore context types) while the `state.matches()` disjunction does that too.

Overall I'd prefer to err on the side of more type safety here if that would be alright with you?

tonyghiani replied:
It makes sense yep, this comment was based on the other thought of extracting the states into an enumeration, so in case of changing a state, the enum would have checked for it. Anyway, shaping the implementation in favour of type safety is totally worth it :)

tonyghiani reviewed (commented):
First of all, great job on the state machines definition, I like the flow you are implementing for handling the state 👏

Reading the code implementation, I've got some questions/suggestions I'd like to talk about:
- The convention when defining XState events is to use the CONST_CASE for the events types. I found it difficult to navigate the machine configuration since the events are all defined in camelCase. 
Also, I noticed some event types have the same name as states (e.g. `loadingLogViewFailed`), I think would be better to have unique names when dealing with states and events to differentiate them :)
- Would be nice to extract the machines' states and event types into enums, it could be useful to have a shared signature for the config definition, events triggering and state matching, over the fact it helps us have an explicit overview of what states/events the machine manages when reading the code.
- As you reported, the `useSelector` hook give great advantages in term of performance to avoid unnecessary re-renders. However, in my understanding, the `useLogView` calls this hook to match against almost all, the machine states, triggering anyway a rerender since we can only be in a specific state and making the usage of useSelector here a de-optimization. I'd like to see how you approached this part and verify if I'm wrong on this, I'm just wondering if it really saves some renders in a complete user-journey or if it can be removed and match all the states using the state object once.

tonyghiani commented on x-pack/plugins/infra/public/observability_logs/log_stream_page/state/src/state_machine.ts:
```diff
@@ -0,0 +1,127 @@
+} from './types';
+
+export const createPureLogStreamPageStateMachine = (initialContext: LogStreamPageContext = {}) =>
+  /** @xstate-layout N4IgpgJg5mDOIC5QBsD2UDKAXATmAhgLYAK+M2+WYAdAK4B2Alk1o-sowF6QDEa+EZlAAy6AGqMwAdwo4qEANoAGALqJQAB1SxGrVPXUgAHogC0ARmoA2ACwAOGzYCsATgDMNt3adWA7L4AaEABPRAAmP2obMPslOzCw3zdfcysAXzSgtExcAhIyMAoqOiYWNg5uCD5UASFRKAlpADF8RmRIZTUkEC0dPQNukwRTMPMw6zc3c1SE1yclGyDQhDcrK2srFycnG18I1aVUjKz0bDwiUnIsShoGZl1yrl5+QXoRcUkZWgBjb7BIDqqQy9B76QxDFK+ahKfzbSZuOLmbZLRBjJRRMIeLzmOIuHG+GzHEDZM55S6Fa7FO5l ...
```
**question**: What does this long token represent?

weltenwort replied:
This is the way the XState Editor stores its layout state. So everyone can just open this up in the editor the VSCode extension provides and read and modify the state machine visually.

tonyghiani replied:
Absolutely, if we can keep control of how the state machine is represented it's just better 👌

weltenwort commented:
Regarding the naming convention for event types: Looking at the state machine again myself after a few days I have to agree that identical-looking event types and state names make it harder to understand the machine. Even more, searching for one also matches the other which is not helpful.

So I changed the all the `event.type` values to all-caps snake-case, which indeed looks to me like it could be more readable. @Kerry350 @tonyghiani let me know if you agree.

@tonyghiani while renaming the events I also discovered that the `LOG_VIEW_ID_CHANGED` event was never dispatched. I added an effect to remedy that.

tonyghiani commented:
> Regarding the naming convention for event types: Looking at the state machine again myself after a few days I have to agree that identical-looking event types and state names make it harder to understand the machine. Even more, searching for one also matches the other which is not helpful.
> 
> So I changed the all the `event.type` values to all-caps snake-case, which indeed looks to me like it could be more readable. @Kerry350 @tonyghiani let me know if you agree.

Definitely agree, I believe it visually helps to understand if we are talking about events or states at first sight.

> @tonyghiani while renaming the events I also discovered that the `LOG_VIEW_ID_CHANGED` event was never dispatched. I added an effect to remedy that.

👍

Kerry350 commented:
> So I changed the all the event.type values to all-caps snake-case, which indeed looks to me like it could be more readable. @Kerry350 @tonyghiani let me know if you agree.

I agree this is more readable, LGTM 👍

Kerry350 commented:
Thank you for reviewing @tonyghiani, and @weltenwort for responding / making changes.
